freqtrade/rpc/api_server/ws/channel.py

```diff
@@ -0,0 +1,178 @@
+import logging
+from threading import RLock
+from typing import List, Optional, Type
+from uuid import uuid4
+
+from fastapi import WebSocket as FastAPIWebSocket
+
+from freqtrade.rpc.api_server.ws.proxy import WebSocketProxy
+from freqtrade.rpc.api_server.ws.serializer import (HybridJSONWebSocketSerializer,
+                                                    WebSocketSerializer)
+from freqtrade.rpc.api_server.ws.types import WebSocketType
+
+
+logger = logging.getLogger(__name__)
+
+
+class WebSocketChannel:
+    """
+    Object to help facilitate managing a websocket connection
+    """
+
+    def __init__(
+        self,
+        websocket: WebSocketType,
+        channel_id: Optional[str] = None,
+        serializer_cls: Type[WebSocketSerializer] = HybridJSONWebSocketSerializer
+    ):
+
+        self.channel_id = channel_id if channel_id else uuid4().hex[:8]
+
+        # The WebSocket object
+        self._websocket = WebSocketProxy(websocket)
+        # The Serializing class for the WebSocket object
+        self._serializer_cls = serializer_cls
+
+        self._subscriptions: List[str] = []
+
+        # Internal event to signify a closed websocket
+        self._closed = False
+
+        # Wrap the WebSocket in the Serializing class
+        self._wrapped_ws = self._serializer_cls(self._websocket)
+
+    def __repr__(self):
+        return f"WebSocketChannel({self.channel_id}, {self.remote_addr})"
+
+    @property
+    def remote_addr(self):
+        return self._websocket.remote_addr
+
+    async def send(self, data):
+        """
+        Send data on the wrapped websocket
+        """
+        await self._wrapped_ws.send(data)
+
+    async def recv(self):
+        """
+        Receive data on the wrapped websocket
+        """
+        return await self._wrapped_ws.recv()
+
+    async def ping(self):
+        """
+        Ping the websocket
+        """
+        return await self._websocket.ping()
+
+    async def close(self):
+        """
+        Close the WebSocketChannel
+        """
+
+        self._closed = True
```

Comment:
wouldn't we want to call `websocket.close()` here, too?

that method seems implemented in the proxy - but right now i struggle to see where that'd be called.

Reply:
So close is really only for FastAPI websockets, and all it really does is send a close message through the websocket. The way channel manager "closes" channels is by setting them as self._closed = True, then deleting them from the channels dict. This just makes it so we don't send anymore data through those channels, and use no data coming in. I believe Uvicorn actually handles closing the sockets and such.

Reply:
So we don't really need to do anything to "close" the websocket, other than just making sure "send" isn't called on it again which we do by checking if channel.is_closed

Reply:
what if we assume a channel is closed - but the client continues to send data (or request data) ?

Reply:
No it just wouldn't do anything